File: Elev8-Groundstation/MovingAverage.cs
/*
  Elev8 GroundStation

  Copyright 2015 Parallax Inc

  This work is licensed under a Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License.
  http://creativecommons.org/licenses/by-nc-sa/4.0/

  Written by Jason Dorie
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Elev8
{
	public class MovingAverage
	{
		float sampleSum = 0;
		int sampleIndex = 0;
		float[] samples = null;

		public MovingAverage( int NumSamples )
		{
			samples = new float[NumSamples];
		}

		public void AddSample( float s )
		{
			sampleSum -= samples[sampleIndex];
			samples[sampleIndex] = s;
			sampleSum += s;

			sampleIndex = (sampleIndex + 1) % samples.Length;
		}

		public float Value
		{
			get {
				return sampleSum / (float)samples.Length;
			}
		}
	}
}

File: FloatStreamCreator/widgets/movingaverage.cpp
#include "movingaverage.h"
#include <string.h>

MovingAverage::MovingAverage()
{
	sampleIndex = 0;
	sampleCount = 0;
	samples = 0;
	sampleSum = 0;
}

MovingAverage::~MovingAverage()
{
	delete [] samples;
}

void MovingAverage::Init(int NumSamples)
{
	if( samples != 0 ) {
		delete [] samples;
		samples = 0;
	}

	sampleCount = NumSamples;
	samples = new float[ sampleCount ];
	sampleIndex = 0;
	sampleSum = 0;

	memset( samples, 0, sizeof(float) * sampleCount );
}

void MovingAverage::AddSample( float s )
{
	sampleSum -= samples[sampleIndex];
	samples[sampleIndex] = s;
	sampleSum += s;

	sampleIndex = (sampleIndex + 1) % sampleCount;
}
